### PR TITLE
Copy Documents upon Save

### PR DESCRIPTION
**Pull request by suztomo**
Before this PR, `setData` did not copy Map and Lists when saving data. Therefore, modification to the data passed as the argument of `setData`, `add`, and `updateData` resulted in affecting the documents held by cloud_firestore_mocks.

### CODE REVIEW

File: test_driver/cloud_firestore_behaviors.dart [hunk 1]
```diff
@@ -166,11 +266,11 @@ void main() {
 
       final DocumentReference doc = messages.document();
       // updateData requires an existing document
-      await doc.setData({'foo': 'bar'});
-
-      await doc.updateData({
-        'nested.data.message': 'old value1',
-        'nested.data.unaffected_field': 'old value2',
+      await doc.setData({
```
Comment (suztomo) [inline test_driver/cloud_firestore_behaviors.dart]: When I wrote the old code, I thought nested document is only for `updateData`. Actually `setData` can take nested documents.

File: test/cloud_firestore_mocks_test.dart [hunk 2]
```diff
@@ -470,6 +470,93 @@ void main() {
...
+    // 3: updateData
+    final DocumentReference document3 = messages.document();
+    await document3.setData({});
+    await document3.updateData({
+      'array': array,
+      'map': map,
+    });
+
+    // The following modifications have no effect the data in Firestore
+    array.add(3);
```
Comment (suztomo) [inline test/cloud_firestore_mocks_test.dart]: This test case is the motivation of this PR.

File: test_driver/cloud_firestore_behaviors.dart [hunk 3]
```diff
@@ -91,14 +91,114 @@ void main() {
...
+      expect(result.data['message'], 'hello firestore');
+      final map1 = result.data['nested1'] as Map<String, dynamic>;
+      expect(map1['field2'], 2);
+      final map2 = map1['nested2'] as Map<String, dynamic>;
+      expect(map2['field3'], 3);
+      final map3 = map2['nested3'] as Map<String, dynamic>;
+      expect(map3['field4'], 4);
+    });
+
+    ftest('Documents should be saved saparately', (firestore) async {
```
Comment (suztomo) [inline test_driver/cloud_firestore_behaviors.dart]: This is pretty much the same as the one in cloud_firestore_mocks_test.dart.

File: lib/src/mock_document_reference.dart [hunk 4]
```diff
@@ -119,20 +122,19 @@ class MockDocumentReference extends Mock implements DocumentReference {
...
-    fromMap.forEach((key, value) {
-      if (value is Map<String, dynamic>) {
-        toMap[key] = _deepCopy(value);
-      } else if (value is List) {
-        toMap[key] = List.from(value);
-      } else {
-        toMap[key] = value;
-      }
+/// Copies data replicating its inner Maps and Lists.
+dynamic deepCopy(dynamic fromData) {
```
Comment (atn832) [inline lib/src/mock_document_reference.dart:128]: Maybe move this to util.dart?

File: test_driver/cloud_firestore_behaviors.dart (hunk 3 above)
Comment (atn832) [reply test_driver/cloud_firestore_behaviors.dart:117]: (btw, you left a typo)

File: lib/src/mock_document_reference.dart [hunk 5]
```diff
@@ -119,20 +122,19 @@ class MockDocumentReference extends Mock implements DocumentReference {
...
+  if (fromData is Map<String, dynamic>) {
+    final toMap = Map<String, dynamic>.from(fromData);
+    toMap.forEach((key, value) {
+      toMap[key] = deepCopy(value);
     });
-
     return toMap;
+  } else if (fromData is List) {
+    return fromData.map((v) => deepCopy(v)).toList();
```
Comment (atn832) [inline lib/src/mock_document_reference.dart:136]: Can you write it as `fromData.map(deepCopy).toList();`?

File: test_driver/cloud_firestore_behaviors.dart (hunk 3 above)
Comment (suztomo) [reply test_driver/cloud_firestore_behaviors.dart]: Fixed.

File: lib/src/mock_document_reference.dart (hunk 5 above)
Comment (suztomo) [reply lib/src/mock_document_reference.dart]: Nice. Updated.

File: lib/src/mock_document_reference.dart (hunk 4 above)
Comment (suztomo) [reply lib/src/mock_document_reference.dart]: Moved this and validateDocumentValue to util.dart.

**Comment by atn832** [comment]
I've just published the PRs you've written this past month to https://pub.dev/packages/cloud_firestore_mocks#042

Thank you!

**Comment by suztomo** [comment]
@atn832 Thank you for the release.

**Pull request merged**